# KPMG Virtual Internship Project

## Task 1: Data Quality Assessment
### Assess the data quality and prepare for analysis

KPMG were provided with the following datasets:
 - Customer Demographic 
 - Customer Addresses
 - Transaction data in the past three months

In [1]:
#importing libraries
import pandas as pd

## Read data

In [2]:
# function to create dataframes from excel sheets
def create_df(filename, sheet_name):
    return pd.read_excel(filename, sheet_name=sheet_name, header=1)

In [5]:
# excel workbook
file = '../data/KPMG_VI_New_raw_data_update_final.xlsx'

## Transactions

In [6]:
# transactions dataframe
transactions = create_df(file, 'Transactions')


In [7]:
transactions.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

In [8]:
transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

In [9]:
# checking for null values
transactions.isna().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

There is missing values in the following columns:
- `online_order` : 360
- `brand`: 197
- `product_line`: 197
- `product_class`: 197
- `product_size`: 197
- `standard_cost` : 197
- `product_first_sold_date` : 197

In [10]:
# function to calculate missing percentage
def miss_percent(df, column):
    miss = (df[column].isna().sum() / len(df[column]))*100
    return print(f'There is {miss:.2f} percent of missing values in {column}')

In [11]:
# understanding size of missing values
miss_percent(transactions, 'online_order')
miss_percent(transactions, 'brand')
miss_percent(transactions, 'product_line')
miss_percent(transactions, 'product_class')
miss_percent(transactions, 'product_size')
miss_percent(transactions, 'standard_cost')
miss_percent(transactions, 'product_first_sold_date')

There is 1.80 percent of missing values in online_order
There is 0.98 percent of missing values in brand
There is 0.98 percent of missing values in product_line
There is 0.98 percent of missing values in product_class
There is 0.98 percent of missing values in product_size
There is 0.98 percent of missing values in standard_cost
There is 0.98 percent of missing values in product_first_sold_date


The missing values are not big compared to the total size of the values.
Depending on the analysis, they will be treated accordingly.

In [12]:
# checking for duplicate values
transactions.duplicated().sum()

0

There are no duplicate values.

### Exploring columns

In [13]:
columns = list(transactions.columns)

In [14]:
# function to get value counts
def count_values(df, columns):
    for col in columns:
        print(f"Value counts for {col}:")
        print(df[col].value_counts())
        print("\n")

In [15]:
count_values(transactions, columns)

Value counts for transaction_id:
2047     1
10912    1
12947    1
2708     1
661      1
        ..
7481     1
1338     1
3387     1
13628    1
2049     1
Name: transaction_id, Length: 20000, dtype: int64


Value counts for product_id:
0      1378
3       354
1       311
35      268
38      267
       ... 
71      137
8       136
16      136
100     130
47      121
Name: product_id, Length: 101, dtype: int64


Value counts for customer_id:
2183    14
1068    14
2476    14
2072    13
1672    13
        ..
2395     1
1569     1
1203     1
2271     1
1544     1
Name: customer_id, Length: 3494, dtype: int64


Value counts for transaction_date:
2017-02-14    82
2017-08-18    82
2017-10-15    76
2017-01-31    73
2017-12-19    71
              ..
2017-01-12    38
2017-12-07    37
2017-03-29    36
2017-09-25    35
2017-10-19    32
Name: transaction_date, Length: 364, dtype: int64


Value counts for online_order:
1.0    9829
0.0    9811
Name: online_order, dtype: int64


Value counts for order_s

`online_order` seems to contain binary features where `1` might denote `YES` and `0` denotes `NO`.<br>
The client will need to explain that column.

In [16]:
# converting column to datetime
transactions['product_first_sold_date'] = pd.to_datetime(transactions['product_first_sold_date'], unit='s')
transactions['product_first_sold_date'].head()

0   1970-01-01 11:27:25
1   1970-01-01 11:35:01
2   1970-01-01 10:06:01
3   1970-01-01 10:02:25
4   1970-01-01 11:43:46
Name: product_first_sold_date, dtype: datetime64[ns]

In [17]:
transactions

transaction_id  product_id  customer_id transaction_date  online_order  \
0                   1           2         2950       2017-02-25           0.0   
1                   2           3         3120       2017-05-21           1.0   
2                   3          37          402       2017-10-16           0.0   
3                   4          88         3135       2017-08-31           0.0   
4                   5          78          787       2017-10-01           1.0   
...               ...         ...          ...              ...           ...   
19995           19996          51         1018       2017-06-24           1.0   
19996           19997          41          127       2017-11-09           1.0   
19997           19998          87         2284       2017-04-14           1.0   
19998           19999           6         2764       2017-07-03           0.0   
19999           20000          11         1144       2017-09-22           1.0   

      order_status           brand product_line product_class product_size  \
0         Approved           Solex     Standard        medium       medium   
1         Approved   Trek Bicycles     Standard        medium        large   
2         Approved      OHM Cycles     Standard           low       medium   
3         Approved  Norco Bicycles     Standard        medium       medium   
4         Approved  Giant Bicycles     Standard        medium        large   
...            ...             ...          ...           ...          ...   
19995     Approved      OHM Cycles     Standard          high       medium   
19996     Approved           Solex         Road        medium       medium   
19997     Approved      OHM Cycles     Standard        medium       medium   
19998     Approved      OHM Cycles     Standard          high       medium   
19999     Approved   Trek Bicycles     Standard        medium        small   

       list_price  standard_cost product_first_sold_date  
0           71.49          53.62     1970-01-01 11:27:25  
1         2091.47         388.92     1970-01-01 11:35:01  
2         1793.43         248.82     1970-01-01 10:06:01  
3         1198.46         381.10     1970-01-01 10:02:25  
4         1765.30         709.48     1970-01-01 11:43:46  
...           ...            ...                     ...  
19995     2005.66        1203.40     1970-01-01 10:30:23  
19996      416.98         312.74     1970-01-01 09:52:40  
19997     1636.90          44.71     1970-01-01 11:13:30  
19998      227.88         136.73     1970-01-01 10:36:56  
19999     1775.81        1580.47     1970-01-01 10:05:34  

[20000 rows x 13 columns]

This column had float data type. The column name states that it is a date hence the conversion<br>
The values however do not seem to be correct because they indicate that the transactions all happened on the same day on different times.

## New Customer List

In [18]:
new_customer_list = create_df(file, 'NewCustomerList')

In [19]:
new_customer_list.head()

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0    Chickie    Brister    Male                                   86   
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   
3     Lucine      Stutt  Female                                   64   
4    Melinda     Hadlee  Female                                   34   

         DOB                   job_title job_industry_category  \
0 1957-07-12             General Manager         Manufacturing   
1 1970-03-22         Structural Engineer              Property   
2 1974-08-28      Senior Cost Accountant    Financial Services   
3 1979-01-28  Account Representative III         Manufacturing   
4 1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car  ...  state    country  \
0      Mass Customer                  N      Yes  ...    QLD  Australia   
1      Mass Customer                  N       No  ...    NSW  Australia   
2  Affluent Customer                  N       No  ...    VIC  Australia   
3  Affluent Customer                  N      Yes  ...    QLD  Australia   
4  Affluent Customer                  N       No  ...    NSW  Australia   

   property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0                   6        0.56      0.7000       0.8750     0.743750   
1                  11        0.89      0.8900       1.1125     0.945625   
2                   5        1.01      1.0100       1.0100     1.010000   
3                   1        0.87      1.0875       1.0875     1.087500   
4                   9        0.52      0.5200       0.6500     0.650000   

   Unnamed: 20  Rank     Value  
0            1     1  1.718750  
1            1     1  1.718750  
2            1     1  1.718750  
3            4     4  1.703125  
4            4     4  1.703125  

[5 rows x 23 columns]

In [20]:
new_customer_list.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   first_name                           1000 non-null   object        
 1   last_name                            971 non-null    object        
 2   gender                               1000 non-null   object        
 3   past_3_years_bike_related_purchases  1000 non-null   int64         
 4   DOB                                  983 non-null    datetime64[ns]
 5   job_title                            894 non-null    object        
 6   job_industry_category                835 non-null    object        
 7   wealth_segment                       1000 non-null   object        
 8   deceased_indicator                   1000 non-null   object        
 9   owns_car                             1000 non-null   object        
 10  tenure       

There are some unnamed columns that without the clients input, we do not know what they signify. They will be dropped for now

In [21]:
# drop the unnamed columns
new_customer_list.drop(['Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18',
       'Unnamed: 19', 'Unnamed: 20'], axis=1, inplace=True)

In [22]:
# check missing values
new_customer_list.isna().sum()

first_name                               0
last_name                               29
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     17
job_title                              106
job_industry_category                  165
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                   0
address                                  0
postcode                                 0
state                                    0
country                                  0
property_valuation                       0
Rank                                     0
Value                                    0
dtype: int64

The following columns contain missing values:
- `last_name` : 29
- `DOB` : 17
- `job_title` : 106
- `job_industry_category` : 165

In [23]:
miss_percent(new_customer_list, 'last_name')
miss_percent(new_customer_list, 'DOB')
miss_percent(new_customer_list, 'job_title')
miss_percent(new_customer_list, 'job_industry_category')

There is 2.90 percent of missing values in last_name
There is 1.70 percent of missing values in DOB
There is 10.60 percent of missing values in job_title
There is 16.50 percent of missing values in job_industry_category


From the percentage seen, appropriate techniques will be used according to the analysis needed.

In [24]:
# check duplicates
new_customer_list.duplicated().sum()

0

There are no duplicated entries

### Exploring columns

In [25]:
columns1 = list(new_customer_list.columns)

In [26]:
count_values(new_customer_list, columns1)

Value counts for first_name:
Mandie      3
Dorian      3
Rozamond    3
Nicol       2
Art         2
           ..
Edin        1
Clevey      1
Cordie      1
Clari       1
Kylynn      1
Name: first_name, Length: 940, dtype: int64


Value counts for last_name:
Eade             2
Shoesmith        2
Crellim          2
Hallt            2
Van den Velde    2
                ..
Howerd           1
Duchant          1
Dwelley          1
McOwan           1
Rochford         1
Name: last_name, Length: 961, dtype: int64


Value counts for gender:
Female    513
Male      470
U          17
Name: gender, dtype: int64


Value counts for past_3_years_bike_related_purchases:
60    20
59    18
70    17
42    17
37    16
      ..
19     5
9      5
92     5
85     4
20     3
Name: past_3_years_bike_related_purchases, Length: 100, dtype: int64


Value counts for DOB:
1961-07-31    2
1994-04-15    2
1987-01-15    2
1951-11-28    2
1979-07-28    2
             ..
1947-04-21    1
1995-10-10    1
1964-08-23    1
196

`gender` has a value `U` that the client needs to explain what it means.

## Customer Demographic

In [27]:
customer_demographic = create_df(file, 'CustomerDemographic')

In [28]:
customer_demographic.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

In [29]:
customer_demographic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

In [30]:
# check null values
customer_demographic.isna().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

The following columns have null values:
- `last_name` : 125
- `DOB` : 87
- `job_title` : 506
- `job_industry_category` : 656
- `default` : 302
- `tenure` : 87

In [31]:
miss_percent(customer_demographic, 'last_name')
miss_percent(customer_demographic, 'DOB')
miss_percent(customer_demographic, 'job_title')
miss_percent(customer_demographic, 'job_industry_category')
miss_percent(customer_demographic, 'default')
miss_percent(customer_demographic, 'tenure')

There is 3.12 percent of missing values in last_name
There is 2.17 percent of missing values in DOB
There is 12.65 percent of missing values in job_title
There is 16.40 percent of missing values in job_industry_category
There is 7.55 percent of missing values in default
There is 2.17 percent of missing values in tenure


The mssing values will be dealt according to the analysis needed due to their sizes.

In [32]:
# check duplicates
customer_demographic.duplicated().sum()

0

There are no duplicates

### Exploring columns

In [33]:
columns2 = list(customer_demographic.columns)

In [34]:
count_values(customer_demographic, columns2)

Value counts for customer_id:
2047    1
657     1
2732    1
681     1
2728    1
       ..
3371    1
1322    1
3367    1
1318    1
2049    1
Name: customer_id, Length: 4000, dtype: int64


Value counts for first_name:
Timmie      5
Max         5
Tobe        5
Urbanus     4
Pail        4
           ..
Purcell     1
Cora        1
Nichols     1
Papageno    1
Jaquelin    1
Name: first_name, Length: 3139, dtype: int64


Value counts for last_name:
Ramsdell     3
Pristnor     3
Vidyapin     2
Scatcher     2
Dyde         2
            ..
McVane       1
Wheatland    1
Perez        1
McCarlie     1
Drache       1
Name: last_name, Length: 3725, dtype: int64


Value counts for gender:
Female    2037
Male      1872
U           88
Femal        1
F            1
M            1
Name: gender, dtype: int64


Value counts for past_3_years_bike_related_purchases:
16    56
19    56
20    54
67    54
2     50
      ..
8     28
86    27
95    27
85    27
92    24
Name: past_3_years_bike_related_purchases, Len

The `default` column has values that are not understandable. Without any input about it from the client, it will be dropped.<br>
The `gender` column has some values that can be combined to `female` or `male`. However there is still one, `U`, which the client will need to explain what it means.

In [35]:
# function to replace a value with another
def substitute(df,col,original_value, sub_value):
    return df[col].replace(original_value, sub_value, inplace=True)

In [36]:
customer_demographic.gender.value_counts()

Female    2037
Male      1872
U           88
Femal        1
F            1
M            1
Name: gender, dtype: int64

In [37]:
# changing values in gender to either male or female
substitute(customer_demographic, 'gender', ['Femal', 'F', 'M'], ['Female', 'Female', 'Male'])

In [38]:
customer_demographic.gender.value_counts()

Female    2039
Male      1873
U           88
Name: gender, dtype: int64

In [39]:
#W dropping the default column
customer_demographic.drop(columns='default', inplace=True)

In [40]:
customer_demographic.columns

Index(['customer_id', 'first_name', 'last_name', 'gender',
       'past_3_years_bike_related_purchases', 'DOB', 'job_title',
       'job_industry_category', 'wealth_segment', 'deceased_indicator',
       'owns_car', 'tenure'],
      dtype='object')

## Customer Address

In [41]:
customer_address = create_df(file, 'CustomerAddress')

In [42]:
customer_address.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [43]:
customer_address.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [44]:
#check null values
customer_address.isna().sum()

customer_id           0
address               0
postcode              0
state                 0
country               0
property_valuation    0
dtype: int64

There are no missing values

In [45]:
# check duplicates
customer_address.duplicated().sum()

0

There are no duplicates

### Exploring the columns

In [46]:
columns3 = list(customer_address.columns)

In [47]:
count_values(customer_address, columns3)

Value counts for customer_id:
2047    1
653     1
2728    1
677     1
2724    1
       ..
3371    1
1322    1
3367    1
1318    1
2049    1
Name: customer_id, Length: 3999, dtype: int64


Value counts for address:
3 Mariners Cove Terrace    2
64 Macpherson Junction     2
3 Talisman Place           2
3723 Bunker Hill Circle    1
3 Golf Course Avenue       1
                          ..
33 Rockefeller Avenue      1
71 Sundown Circle          1
35386 Mockingbird Road     1
3 Hayes Court              1
86 Fairfield Alley         1
Name: address, Length: 3996, dtype: int64


Value counts for postcode:
2170    31
2155    30
2145    30
2153    29
2770    26
        ..
4552     1
4555     1
2485     1
3580     1
4421     1
Name: postcode, Length: 873, dtype: int64


Value counts for state:
NSW                2054
VIC                 939
QLD                 838
New South Wales      86
Victoria             82
Name: state, dtype: int64


Value counts for country:
Australia    3999
Name: country, 

The column `state` has some entries that can be substituted.

In [48]:
substitute(customer_address, 'state', ['New South Wales', 'Victoria'], ['NSW', 'VIC'])

In [49]:
customer_address.state.value_counts()

NSW    2140
VIC    1021
QLD     838
Name: state, dtype: int64

# Conclusions

The datasets have various issues such as:
- Missing values : This is evident in almost all the datasets.
- Incorrect data type : There was a column with a data type that did not reflect the expected data type.
- Different classifiers: What this means is that there were some columns that had values that were classified differently but seemed to mean the same thing. Example, gender had `Femal` which seems to mean `Female`.
- Unknown values: Some columns had values that are not understandable without further explanation from the client.
